funtools/man/man1/funhist.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "funhist 1"
.TH funhist 1 "April 14, 2011" "version 1.4.5" "SAORD Documentation"
.SH "NAME"
funhist \- create a 1D histogram of a column (from a FITS binary table or raw event file) or an image
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBfunhist\fR  [\-n|\-w|\-T] <iname> [column] [[lo:hi:]bins]
.SH "OPTIONS"
.IX Header "OPTIONS"
.Vb 3
\&  \-n    # normalize bin value by the width of each bin
\&  \-w    # specify bin width instead of number of bins in arg3
\&  \-T    # output in rdb/starbase format (tab separators)
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfunhist\fR creates a one-dimensional histogram from the specified
columns of a \s-1FITS\s0 Extension
binary table of a \s-1FITS\s0 file (or from a non-FITS raw event file), or
from a \s-1FITS\s0 image or array, and writes that histogram as an \s-1ASCII\s0
table. Alternatively, the program can perform a 1D projection of one
of the image axes.
.PP
The first argument to the program is required, and specifies the
Funtools file: \s-1FITS\s0 table or image, raw event file, or array.  If
\&\*(L"stdin\*(R" is specified, data are read from the standard input. Use 
Funtools Bracket Notation to specify \s-1FITS\s0
extensions, and filters.
.PP
For a table, the second argument also is required. It specifies the
column to use in generating the histogram.  If the data file is of
type image (or array), the column is optional: if \*(L"x\*(R" (or \*(L"X\*(R"), \*(L"y\*(R"
(or \*(L"Y\*(R") is specified, then a projection is performed over the x
(dim1) or y (dim2) axes, respectively. (That is, this projection will
give the same results as a histogram performed on a table containing
the equivalent x,y event rows.)  If no column name is specified or
\&\*(L"xy\*(R" (or \*(L"\s-1XY\s0\*(R") is specified for the image, then a histogram is
performed on the values contained in the image pixels.
.PP
The argument that follows is optional and specifies the number of bins
to use in creating the histogram and, if desired, the range of bin
values.  For image and table histograms, the range should specify the
min and max data values.  For image histograms on the x and y axes,
the range should specify the min and max image bin values.  If this
argument is omitted, the number of output bins for a table is
calculated either from the \s-1TLMIN/TLMAX\s0 headers values (if these exist
in the table \s-1FITS\s0 header for the specified column) or by going through
the data to calculate the min and max value. For an image, the number
of output bins is calculated either from the \s-1DATAMIN/DATAMAX\s0 header
values, or by going through the data to calculate min and max value.
(Note that this latter calculation might fail if the image cannot be
fit in memory.)  If the data are floating point (table or image) and
the number of bins is not specified, an arbitrary default of 128 is
used.
.PP
For binary table processing, the \fB\-w\fR (bin width) switch can be used
to specify the width of each bin rather than the number of bins. Thus:
.PP
.Vb 1
\&  funhist test.ev pha 1:100:5
.Ve
.PP
means that 5 bins of width 20 are used in the histogram, while:
.PP
.Vb 1

ChangeLog
2016-01-12 waj

	* upgraded TEA to 3.9
	* numerous compiler fixes for 64bit

2008-11-14  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.10 TAGGED FOR RELEASE ****

	* generic/tkTable.c: Set NO_XSETCLIP universally as Tk 8.5/Xft
	seems to ignore XSetClipRectangles.  This may slow things down a
	little, but draws correctly. [Bug 1805350]

	* UPGRADING.txt (removed): no longer need this old v1 upgrade doc.

	* README.txt: remove refs to mac classic stuff, update notes

	* configure, configure.in, tclconfig/tcl.m4: Update to TEA 3.7
	* tests/tkTable.test, unix/tktable.spec, generic/version.h:
	* demos/loadtable.tcl: Update version to 2.10
	* Makefile.in: update dist target, include library/tktable.py

	* demos/tktable.py: note that this version is outdated
	* library/tktable.py (NEW): Added new version of tktable wrapper
	for Python/Tkinter from GPolo.  Not 100% compatible with previous
	version. [Patch 2244167]

	* generic/tkTableTag.c (Table_TagCmd): remove interp->result set
	work-around for old bug as i->result is no longer exposed in 8.6.

	* generic/tkTable.c:     Use safe Tcl_SetObjResult instead of trying
	* generic/tkTableCell.c: to set existing objResult value.
	* generic/tkTableCmds.c: 

2006-07-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* license.txt (bourbon_ware): Be less restrictive in the
	bourbon_ware clause.

2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure.in, configure, tclconfig/tcl.m4: update to TEA 3.5

2006-01-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure: update to TEA 3.4 interim

2005-12-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): remove potential
	undefined behavior warning.

2005-12-02  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure.in, configure: update to TEA 3.4

2005-05-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (html): switch to groff to fix botched html output.
	* doc/tkTable.html: regen [Bug 1198478]
	* doc/tkTable.n: use '' around \\ items.

2005-03-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (AR): use @AR@, improve html doc target
	* configure, configure.in, tclconfig/tcl.m4: TEA 3.2 patch update

	* doc/tkTable.n, doc/kTable.html: doc cleanup

2005-02-01  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in, tclconfig/tcl.m4: update to TEA 3.2

2004-11-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::SelectAll): remove invalid ref
	to HandleType proc. [Bug 1068874] (ade)

2004-07-22  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.9 TAGGED FOR RELEASE ****

	* configure, tclconfig/tcl.m4: update tcl.m4 TEA_PATH_X that adds
	Aqua build recognition and TEA_WINDOWINGSYSTEM variable.

2004-07-20  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/version.h, unix/tktable.spec:    updated for v2.9
	* demos/loadtable.tcl, tests/tkTable.test:

	* Makefile.in, configure.in, configure, tclconfig/tcl.m4: update
	to TEA 3.1, use CFLAGS instead of EXTRA_CFLAGS and bump the
	version number of Tktable to 2.9.

	* generic/tkTable.h: add #include <ctype.h>

	* generic/tkTableCell.c (TableGetCellValue, TableSetCellValue): 
	Do faster cache checks when DATA_CACHE is the only data source.
	In fallback from command that errors to array var, retrieve array
	value immediately on current get|set request.

funtools/funhist.c
/*
 *	Copyright (c) 1999-2003 Smithsonian Astrophysical Observatory
 */

#include <math.h>
#include <funtoolsP.h>
#include <filter.h>
#include <word.h>

static char *dash="----------------------------------------";

extern char *optarg;
extern int optind;

#define MAXROW 8192
static int maxrow=MAXROW;
int cold=' ';

#ifdef ANSI_FUNC
void
GetArgs (int argc, char **argv, int arg,
	 double *tlmin, double *tlmax, double *dval, int *got1, int *got2)
#else
void GetArgs(argc, argv, arg, tlmin, tlmax, dval, got1, got2)
     int argc;
     char **argv;
     int arg;
     double *tlmin;
     double *tlmax;
     double *dval;
     int *got1;
     int *got2;
#endif
{
  int ip=0;
  int n;
  char tbufs[3][SZ_LINE];

  /* process the appropriate argument */
  if( argc > arg ){
    /* collect arguments: bins, or lo:hi, or lo:hi:bins */
    newdtable(",:)");
    for(n=0; (n<3) && word(argv[arg], tbufs[n], &ip); n++)
      ;
    freedtable();
    /* process arguments */
    switch(n){
    case 0:
      break;
    case 1:
      *dval = atof(tbufs[0]);
      break;
    case 2:
      *tlmin = atof(tbufs[0]);
      *tlmax = atof(tbufs[1]);
      *got1 = 1;
      *got2 = 1;
      break;
    case 3:
      *tlmin = atof(tbufs[0]);
      *tlmax = atof(tbufs[1]);
      *got1 = 1;
      *got2 = 1;
      *dval = atof(tbufs[2]);
      break;
    }
  }
}

#ifdef ANSI_FUNC
static void
usage (char *fname)
#else
static void usage(fname)
     char *fname;
#endif
{
  fprintf(stderr,
    "usage: %s [-n|-w|-T] iname [column|x|y|xy] [[lo_edge:hi_edge:]bins]\n",
    fname);
  fprintf(stderr, "optional switches:\n");
  fprintf(stderr, "  -n  # normalize bin value by the width of each bin\n");
  fprintf(stderr, "  -w  # specify bin width instead of number of bins in arg3\n");
  fprintf(stderr, "  -T  # output in rdb/starbase format (tab separators)\n");
  fprintf(stderr, "\n(version: %s)\n", FUN_VERSION);
  exit(1);
}
#ifdef ANSI_FUNC
int 
main (int argc, char **argv)
#else
int
main(argc, argv)
     int argc;
     char **argv;
#endif
{
  int c;
  int i, j;
  int in, ix, iy;
  int arg, args;
  int idx;
  int got;
  int nrow;
  int rsize;
  int bitpix;
  int x0, y0;
  int x1, y1;
  int block;
  int dim1, dim2;
  int dtype;
  int type;
  int lasty=0;
  int obins=0;
  int got1=0, got2=0;
  int gotimage=0;
  int doproj = 0;
  int donorm = 0;
  int dowidth = 0;
  int nmask=0;
  int nreg=0;
  char *s;
  char *iname;
  char *col;
  char *filtstr=NULL;
  char *bincols=NULL;
  char tbuf[SZ_LINE];
  double tlmin=0, tlmax=0;
  double tlwidth, tlinc;
  double lo, hi;
  double dval=0.0;
  double *buf=NULL, *rbuf=NULL, *rptr=NULL;
  Filter filter;
  FilterMask masks=NULL;
  FITSHead header=NULL;
  Fun fun=NULL;

  /* exit on gio errors */
  if( !getenv("GERROR")  )
    setgerror(2);

  /* we want the args in the same order in which they arrived, and
     gnu getopt sometimes changes things without this */
  putenv("POSIXLY_CORRECT=true");

  /* process switch arguments */
  while ((c = getopt(argc, argv, "nwT")) != -1){
    switch(c){
    case 'n':
      donorm = 1;
      break;
    case 'w':
      dowidth = 1;
      break;
    case 'T':
      cold = '\t';
      break;
    }
  }
  
  /* get maxrow,if user-specified */
  if( (s=getenv("FUN_MAXROW")) != NULL )
    maxrow = atoi(s);

  /* process arguments */
  /* check for required arguments */
  args = argc - optind;
  if( args < 1 ) usage(argv[0]);
  iname = argv[optind+0];
  /* open the input FITS file */
  if( !(fun = FunOpen(iname, "r", NULL)) )
    gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	   iname);
  /* make sure we have the right type of input file */
  FunInfoGet(fun, FUN_TYPE, &type, 0);

  /* process the different types */
  switch(type){
  case FUN_TABLE:
  case FUN_EVENTS:
    if( args < 2 ) usage(argv[0]);
    /* get column */
    col = argv[optind+1];
    /* make sure we can find the specified column */
    if( !(idx = FunColumnLookup(fun, col, 0, NULL, 
				&dtype, NULL, NULL, NULL, NULL)) )
      gerror(stderr, "can't find specified column: %s\n", col);
    /* get tlmin/tlmax, etc. */
    GetArgs(argc, argv, optind+2, &tlmin, &tlmax, &dval, &got1, &got2);
    if( dowidth && (dval<=0) )
	gerror(stderr, "-w requires a positive bin width value\n");
    else
      obins = (int)dval;
    /* select column to read */
    FunColumnSelect(fun, sizeof(double), NULL, col, "D", "r", 0, NULL);
    /* for fp data, we cannot determine obins, so choose an arbitrary value */
    if( !dowidth && ((dtype == 'D') || (dtype == 'E')) && (obins <= 0) )
      obins = 128;
    /* read specified column for all rows into a double array */
    rsize = maxrow;
    rbuf = xmalloc(rsize*sizeof(double));
    rptr = rbuf;
    nrow = 0;
    while( 1 ){
      FunTableRowGet(fun, rptr, maxrow, NULL, &got);
      if( !got ) break;
      rsize += maxrow;
      rbuf = xrealloc(rbuf, rsize*sizeof(double));
      nrow += got;
      rptr = rbuf + nrow;
    }
    /* get tlmin value */
    if( !got1 ){
      snprintf(tbuf, SZ_LINE, "TLMIN%d", idx);
      tlmin = FunParamGetd(fun, tbuf, 0, 0.0, &got1);
    }
    /* get tlmax value */
    if( !got2 ){
      snprintf(tbuf, SZ_LINE, "TLMAX%d", idx);
      tlmax = FunParamGetd(fun, tbuf, 0, 0.0, &got2);
    }
    /* if we do not have tlmin or tlmax, we need to calculate them now */
    if( !got1 || !got2 ){
      if( !got1 ) tlmin = rbuf[0];
      if( !got2 ) tlmax = rbuf[0];
      for(i=0; i<nrow; i++){
	if( !got1 && (rbuf[i] < tlmin) ) tlmin = rbuf[i];
	if( !got2 && (rbuf[i] > tlmax) ) tlmax = rbuf[i];
      }
      /* add 1/2 bin to ends of float tlmin and tlmax */
      if( !got1 && ((dtype == 'D') || (dtype == 'E')) ) tlmin -= 0.5;
      if( !got2 && ((dtype == 'D') || (dtype == 'E')) ) tlmax += 0.5;
    }
    /* tlmin/tlmax are displayed as edges, even with int data */
    if( (dtype != 'D') && (dtype != 'E') ){
      tlmin -= 0.5;
      tlmax += 0.5;
    }
    /* width was specified instead of number of bins */
    if( dowidth ){
      tlinc = dval;
      obins = (int)((tlmax - tlmin)/tlinc);
      if( obins <= 0 )
	gerror(stderr, "calculated number of bins (%d) must be positive\n",
	       obins);
      tlwidth = (double)obins*tlinc;
      if( tlwidth <= 0 )
	gerror(stderr, "calculated bin width (%.8f) must be positive\n",
	       tlwidth);
      tlmax = tlmin+tlwidth;
    }
    else{
      /* get total number of input bins */
      tlwidth = tlmax - tlmin;
      /* calculate the number of output bins */
      if( obins <= 0 )
	obins = (int)tlwidth;
      /* this is the increment between each bin */
      tlinc = tlwidth/obins;
      if( tlwidth <= 0 )
	gerror(stderr, "bin width (%.8f) must be positive\n", tlwidth);
    }

    /* allocate bin buffer */
    if( !(buf=xcalloc(sizeof(double), obins)) )
      gerror(stderr, "can't allocate image buffer\n");
    /* bin up all rows */
    for(i=0; i<nrow; i++){
      if( (rbuf[i]<tlmin) || (rbuf[i]>tlmax) )
	continue;
      idx = (int)(((rbuf[i] - tlmin)/tlwidth)*obins);
      if( (idx<0) || (idx>=obins) )
	continue;
      buf[idx] += 1;
    }
    /* output header */
    fprintf(stdout, "# data file:\t\t%s\n", iname);
    fprintf(stdout, "# column:\t\t%s\n", col);
    fprintf(stdout, "# min,max,bins:\t\t%g %g %d\n", tlmin, tlmax, obins);
    if( dowidth )
      fprintf(stdout, "# user-specified width: %.8f\n", tlinc);
    if( donorm )
      fprintf(stdout, "# width normalization applied: %.8f\n", tlinc);
    fprintf(stdout, "\n");
    /* output data */
    if( donorm ){
      fprintf(stdout,
	      "%6s%c%21s%c%21s%c%21s\n", 
	      "bin", cold, "value", cold, "lo_edge", cold, "hi_edge");
      fprintf(stdout,
	      "%6.6s%c%21.21s%c%21.21s%c%21.21s\n",
	      dash, cold, dash, cold, dash, cold, dash);
    }
    else{
      fprintf(stdout,
	      "%6s%c%9s%c%21s%c%21s\n", 
	      "bin", cold, "value", cold, "lo_edge", cold, "hi_edge");
      fprintf(stdout,
	      "%6.6s%c%9.9s%c%21.21s%c%21.21s\n",
	      dash, cold, dash, cold, dash, cold, dash);
    }
    for(i=0; i<obins; i++){
      lo = tlmin+(i*tlinc);
      hi = tlmin+((i+1)*tlinc);
      if( donorm ){
	fprintf(stdout, "%6d%c%21.8f%c%21.8f%c%21.8f\n", 
		i+1, cold, buf[i]/tlinc, cold, lo, cold, hi);
      }
      else{
	fprintf(stdout, "%6d%c%9d%c%21.8f%c%21.8f\n", 
		i+1, cold, (int)buf[i], cold, lo, cold, hi);
      }
    }
    /* free up buffers */
    if( buf )  xfree(buf);
    if( rbuf ) xfree(rbuf);
    break;

  case FUN_IMAGE:
  case FUN_ARRAY:
    /* -w only useful for tabular data */
    if( dowidth )
      gerror(stderr, "specifying bin width (-w) not supported with images\n");
    /* determine if we are doing a proj on a single column or a histogram
       on the array values */
    arg = optind+1;
    if( args >=2 ){
      col = argv[optind+1];
      cluc(col);
      if( !strcasecmp(col, "X") || !strcasecmp(col, "Y") ){
	doproj = 1;
	arg = optind+2;
      }
      else if( !strcasecmp(col, "XY") || !strcasecmp(col, "YX") ){
	arg = optind+2;
      }
      else
	gerror(stderr, "can't find specified column: %s\n", col);
    }
    /* projection of a single axis */
    if( doproj ){
      /* get tlmin/tlmax, etc. */
      GetArgs(argc, argv, arg, &tlmin, &tlmax, &dval, &got1, &got2);
      obins = (int)dval;
      FunInfoGet(fun,
		 FUN_SECT_BLOCK, &block,
		 FUN_SECT_X0, &x0, FUN_SECT_Y0, &y0,
		 FUN_SECT_DIM1, &dim1, FUN_SECT_DIM2, &dim2,
		 0);
      /* set dim2 to 1 for a 1D image */
      if( dim2 == 0 ) dim2 = 1;
      /* get tlmin/tlmax values, etc. */
      if( tlmin <=0 )
	tlmin = 1;
      if( tlmax <=0 ){
	if( *col == 'X' )
	  tlmax = dim1;
	else
	  tlmax = dim2;
      }
      /* get total number of input bins */
      tlwidth = tlmax - tlmin + 1;
      /* calculate the number of output bins */
      if( obins <= 0 ){
	/* use input bin number as default */
	obins = (int)tlwidth;
      }
      tlinc = tlwidth/obins;
      /* allocate bin buffer */
      if( !(buf=xcalloc(sizeof(double), obins)) )
	gerror(stderr, "can't allocate image buffer\n");
      if( *col == 'X' ){
	for(i=1; i<=dim2; i++){
	  if( !(rbuf = FunImageRowGet(fun, rbuf, i, i, "bitpix=-64")) )
	    gerror(stderr, "can't FunImageRowGet: %s\n", iname);
	  for(j=(int)tlmin; j<=(int)tlmax; j++){
	    idx = (int)(((j-tlmin)/tlwidth)*obins);
	    if( (idx<0) || (idx>=obins) )
	      continue;
	    buf[idx] += rbuf[j-1];
	  }
	}
      }
      else{
	for(i=(int)tlmin; i<=(int)tlmax; i++){
	  if( !(rbuf = FunImageRowGet(fun, rbuf, i, i, "bitpix=-64")) )
	    gerror(stderr, "can't FunImageRowGet: %s\n", iname);
	  idx = (int)(((i-tlmin)/tlwidth)*obins);
	  if( (idx<0) || (idx>=obins) )
	    continue;
	  for(j=0; j<dim1; j++)
	    buf[idx] += rbuf[j];
	}
      }
      /* output header */
      fprintf(stdout, "# data file:\t\t%s\n", iname);
      fprintf(stdout, "# column:\t\t%s\n", col);
      fprintf(stdout, "# min,max,bins:\t\t%g %g %d\n", tlmin, tlmax, obins);
      fprintf(stdout, "\n");
      /* output data */
      fprintf(stdout,
	      "%6s%c%21s%c%21s%c%21s\n",
	      "bin", cold, "value", cold, "lo_bin", cold, "hi_bin");
      fprintf(stdout,
	      "%6.6s%c%21.21s%c%21.21s%c%21.21s\n",
	      dash, cold, dash, cold, dash, cold, dash);
      for(i=0; i<obins; i++){
	  fprintf(stdout, "%6d%c%21.8f%c%21.8f%c%21.8f\n", 
		  i+1, cold, buf[i], cold,
		  ceil(tlmin+(i*tlinc)), cold, ceil(tlmin+((i+1)*tlinc))-1);
      }
      /* free up buffers */
      if( buf )  xfree(buf);
      if( rbuf ) xfree(rbuf);
    }
    /* histogram of the values in the image */
    else{
      /* get tlmin/tlmax, etc. */
      GetArgs(argc, argv, arg, &tlmin, &tlmax, &dval, &got1, &got2);
      obins = (int)dval;
      FunInfoGet(fun,
		 FUN_FILTER, &filtstr, FUN_HEADER, &header,
		 FUN_SECT_X0, &x0, FUN_SECT_Y0, &y0,
		 FUN_SECT_X1, &x1, FUN_SECT_Y1, &y1,
		 FUN_SECT_BLOCK, &block, FUN_BITPIX, &bitpix,
		 FUN_BINCOLS, &bincols,
		 FUN_SECT_DIM1, &dim1, FUN_SECT_DIM2, &dim2,
		 0);
      /* set dim2 to 1 for a 1D image */
      if( dim2 == 0 ) dim2 = 1;
      /* get filter masks, if necessary */
      if( filtstr && *filtstr ){
	/* make up filter plist string */
	strcpy(tbuf, "type=image");
	/* add the binning key */
	if( bincols ){
	  strcat(tbuf, ",");
	  strcat(tbuf, bincols);
	}
	filter = FilterOpen(header, filtstr, tbuf);
	if( filter && (filter != NOFILTER) ){
	  /* retrieve region mask segments */
	  nmask = FilterImage(filter, x0, x1, y0, y1, block, &masks, &nreg);
	}
	FilterClose(filter);
      }
      /* if min/max not given on command line, look in data file */
      if( !got1 )
	tlmin=FunParamGetd(fun, "DATAMIN", 0, 0.0, &got1);
      if( !got2 )
	tlmax=FunParamGetd(fun, "DATAMAX", 0, 0.0, &got2);
      /* for image, we must have min/max values */
      if( !got1 || !got2 ){
	if( (rbuf = FunImageGet(fun, rbuf, "bitpix=-64")) ){
	  if( nmask ){
	    tlmin = MAXDBL;
	    tlmax = MINDBL;
	    /* loop through mask segments */
	    for(in=0; in<nmask; in++){
	      iy=masks[in].y;
	      /* process each mask segment having the same y value */
	      while(1){
		/* process columns in the mask */
		for(ix=masks[in].xstart; ix<=masks[in].xstop; ix++){
		  i = ((iy-1)*dim1)+(ix-1);
		  if( rbuf[i] < tlmin )
		    tlmin = rbuf[i];
		  if( rbuf[i] > tlmax )
		    tlmax = rbuf[i];
		}
		/* if another mask segment has the same y, keep going */
		if( ((in+1)<nmask) && (masks[in+1].y==iy) )
		  in++;
		else
		  break;
	      }
	    }
	  }
	  else{
	    tlmin = rbuf[0];
	    tlmax = rbuf[0];
	    for(i=0; i<dim1*dim2; i++){
	      if( rbuf[i] < tlmin )
		tlmin = rbuf[i];
	      if( rbuf[i] > tlmax )
		tlmax = rbuf[i];
	    }
	  }
	  gotimage = 1;
	}
	else{
	  gerror(stderr, 
	 "please supply min:max:bins (file is lacking DATAMIN/DATAMAX)\n");
	}
      }
      /* get total number of input bins */
      if( bitpix >= 0 )
	tlwidth = tlmax - tlmin + 1;
      else
	tlwidth = tlmax - tlmin;
      /* calculate the number of output bins */
      if( obins <= 0 ){
	/* use input bin number as default */
	if( bitpix >= 0 )
	  obins = (int)tlwidth;
	/* but for floating point data, obins must a valid int */
	else
	  obins = 128;
      }
      /* this is the increment between each bin */
      tlinc = tlwidth/obins;
      /* allocate bin buffer */
      if( !(buf=xcalloc(sizeof(double), obins)) )
	gerror(stderr, "can't allocate image buffer\n");
      /* bin up all rows */
      if( gotimage ){
	if( nmask ){
	  /* loop through mask segments */
	  for(in=0; in<nmask; in++){
	    iy=masks[in].y;
	    /* process each mask segment having the same y value */
	    while(1){
	      /* process columns in the mask */
	      for(ix=masks[in].xstart; ix<=masks[in].xstop; ix++){
		j = ((iy-1)*dim1)+(ix-1);
		if( (rbuf[j]<tlmin) || (rbuf[j]>tlmax) )
		  continue;
		idx = (int)(((rbuf[j] - tlmin)/tlwidth)*obins);
		if( (idx<0) || (idx>=obins) )
		  continue;
		buf[idx] += 1;
	      }
	      /* if another mask segment has the same y, keep going */
	      if( ((in+1)<nmask) && (masks[in+1].y==iy) )
		in++;
	      else
		break;
	    }
	  }
	}
	else{
	  for(j=0; j<dim1*dim2; j++){
	    if( (rbuf[j]<tlmin) || (rbuf[j]>tlmax) )
	      continue;
	    idx = (int)(((rbuf[j] - tlmin)/tlwidth)*obins);
	    if( (idx<0) || (idx>=obins) )
	      continue;
	    buf[idx] += 1;
	  }
	}
      }
      else{
	if( nmask ){
	  /* loop through mask segments */
	  for(in=0; in<nmask; in++){
	    iy=masks[in].y;
	    if( iy != lasty ){
	      if( !(rbuf = FunImageRowGet(fun, rbuf, iy, iy, "bitpix=-64")) )
		gerror(stderr, "can't FunImageRowGet: %d %s\n", iy, iname);
	      lasty = iy;
	    }
	    /* process each mask segment having the same y value */
	    while(1){
	      /* process columns in the mask */
	      for(ix=masks[in].xstart; ix<=masks[in].xstop; ix++){
		j = ((iy-1)*dim1)+ix-1;
		if( (rbuf[j]<tlmin) || (rbuf[j]>tlmax) )
		  continue;
		idx = (int)(((rbuf[j] - tlmin)/tlwidth)*obins);
		if( (idx<0) || (idx>=obins) )
		  continue;
		buf[idx] += 1;
	      }
	      /* if another mask segment has the same y, keep going */
	      if( ((in+1)<nmask) && (masks[in+1].y==iy) )
		in++;
	      else
		break;
	    }
	  }
	}
	else{
	  for(i=0; i<dim2; i++){
	    if( !(rbuf = FunImageRowGet(fun, rbuf, i+1, i+1, "bitpix=-64")) )
	      gerror(stderr, "can't FunImageRowGet: %d %s\n", i+1, iname);
	    /* bin up all rows */
	    for(j=0; j<dim1; j++){
	      if( (rbuf[j]<tlmin) || (rbuf[j]>tlmax) )
		continue;
	      idx = (int)(((rbuf[j] - tlmin)/tlwidth)*obins);
	      if( (idx<0) || (idx>=obins) )
		continue;
	      buf[idx] += 1;
	    }
	  }
	}
      }
      /* output header */
      fprintf(stdout, "# data file:\t\t%s\n", iname);
      fprintf(stdout, "# min,max,bins:\t\t%g %g %d\n", tlmin, tlmax, obins);
      fprintf(stdout, "\n");
      /* output data */
      fprintf(stdout,
	      "%6s%c%21s%c%21s%c%21s\n",
	      "bin", cold, "value", cold, "lo_val", cold, "hi_val");
      fprintf(stdout,
	      "%6.6s%c%21.21s%c%21.21s%c%21.21s\n",
	      dash, cold, dash, cold, dash, cold, dash);
      for(i=0; i<obins; i++){
	if( bitpix >= 0 )
	  fprintf(stdout, "%6d%c%21.8f%c%21.8f%c%21.8f\n", 
		  i+1, cold, buf[i], cold,
		  ceil(tlmin+(i*tlinc)), cold, ceil(tlmin+((i+1)*tlinc))-1);
	else
	  fprintf(stdout, "%6d%c%21.8f%c%21.8f%c%21.8f\n", 
		  i+1, cold, buf[i], cold,
		  tlmin+(i*tlinc), cold, tlmin+((i+1)*tlinc));
      }
      /* free up buffers */
      if( buf )   xfree(buf);
      if( rbuf )  xfree(rbuf);
      if( masks ) xfree(masks);
    }
    break;
  }


  /* close files */
  FunClose(fun);

  return(0);
}
